clang/include/clang/Basic/DiagnosticSemaKinds.td

```diff
@@ -12199,6 +12199,9 @@ def err_sycl_kernel_name_type : Error<
 def err_sycl_kernel_name_conflict : Error<
   "'sycl_kernel_entry_point' kernel name argument conflicts with a previous"
   " declaration">;
+def err_builtin_invalid_argument_count : Error<"builtin requires exactly 1 argument">;
+def err_builtin_invalid_argument : Error<"invalid argument; argument must be a class or struct type"
+                                         " with a member type alias named 'type'">;
```

Comment:
I took inspiration from some existing diagnostics with "argument must be a xyz" wording. I personally prefer that to "expected a...". I have no strong feelings though and so I have made this modification. 

I did not specify the type via %0 in the diagnostic because we have the ^ pointing to the problematic type when we emit the diagnostic. Way back, someone told me to not include unnecessary information in diagnostics to keep it short, if we are already pointing to the problematic type.

Reply:
I'm fine with the "argument must be a xyz" wording; that is consistent with other diagnostics too.

My motivation for including the type in the diagnostic is because the carat won't actually point to the type; I expect it will point to the expression that produces an object of the type. Given this:
```
template<typename KernelName>
struct kernel_id_t {
  using type = KernelName;
};
template<typename KernelName>
void f() {
  __builtin_sycl_kernel_name(kernel_id_t<KernelName>());
}
```

I would expect a diagnostic to look something like:
```
invalid argument; argument must be a class or structure with a member typedef or type alias named 'type'
  __builtin_sycl_kernel_name(kernel_id_t<KernelName>());
                             ^^^^^^^^^^^^^^^^^^^^^^^^^
```

Our expectation is that the expression will readily identify the type in expected usage as in the above case, but that isn't guaranteed. For example, the expression could call some function (I don't know of any good reason to do so, but style preferences could lead someone to do so).